src/dynimage_serde.rs

```diff
@@ -0,0 +1,182 @@
+                    where
+                        E: de::Error,
+                    {
+                        match value {
+                            "dtype" => Ok(Field::Dtype),
+                            "data" => Ok(Field::Data),
+                            _ => Err(de::Error::unknown_field(value, FIELDS)),
+                        }
+                    }
+                }
+
+                deserializer.deserialize_identifier(FieldVisitor)
+            }
+        }
+
+        struct SerialBufferVisitor;
+        struct DynamicSerialImage {
+            dtype: String,
+            data: String,
+        }
+
+        impl<'de> Visitor<'de> for SerialBufferVisitor {
+            type Value = DynamicSerialImage;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("struct DynamicSerialImage")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: de::SeqAccess<'de>,
+            {
+                let dtype: String = seq
+                    .next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(0, &self))?;
+                let data: String = seq
+                    .next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(1, &self))?;
+                Ok(DynamicSerialImage { dtype, data })
+            }
+
+            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+            where
+                A: de::MapAccess<'de>,
+            {
+                let mut data: Option<String> = None;
+                let mut dtype: Option<String> = None;
+
+                while let Some(key) = map.next_key()? {
+                    match key {
+                        Field::Dtype => {
+                            if dtype.is_some() {
+                                return Err(de::Error::duplicate_field("dtype"));
+                            }
+                            dtype = Some(map.next_value()?);
+                        }
+                        Field::Data => {
+                            if data.is_some() {
+                                return Err(de::Error::duplicate_field("data"));
+                            }
+                            data = Some(map.next_value()?);
+                        }
+                    }
+                }
+
+                let dtype = dtype.ok_or_else(|| de::Error::missing_field("dtype"))?;
+                let data = data.ok_or_else(|| de::Error::missing_field("data"))?;
+
+                Ok(DynamicSerialImage { dtype, data })
+            }
+        }
+
+        const FIELDS: &[&str] = &["dtype", "data"];
+        let res =
+            deserializer.deserialize_struct("DynamicSerialImage", FIELDS, SerialBufferVisitor)?;
+
+        let mut imgdata = STANDARD_NO_PAD.decode(res.data.as_bytes()).map_err(|e| {
```

Comment:
I'd rather skip the base64 coding step. The efficient serialization formats like bincode actually get worse sending base64 strings rather than the raw bytes